### PR TITLE
Reduce network burden of the hunger system

### PR DESCRIPTION
**Pull request by Centronias**
## About the PR
Replace `HungerComponent`'s `CurrentHunger` with `HungerSystem.GetHunger`, which calculates the current hunger value from an initial value, a time when that initial value was set, and a constant decay rate.

See also #33403 

## Why / Balance
#28070

## Technical details
* `HungerComponent` no longer has `CurrentHunger`, and instead has `LastAuthoritativeHungerValue` and `LastAuthoritativeHungerChangeTime`
* Replaced all usages of `HungerComponent.CurrentHunger` with `HungerSystem.GetHunger`, which calculates the current hunger value by predicting the decay since the last authoritative update:
  * `LastAuthoritativeHungerValue - (IGameTiming.CurrTime - LastAuthoritativeHungerChangeTime) * ActualDecayRate`
* `HungerSystem.Update` now only looks to update the current hunger threshold (because there's no value to update)
* Added `HungerSystem.SetAuthoritativeHungerValue`, which sets `LastAuthoritativeHungerValue` and `LastAuthoritativeHungerChangeTime` and dirties the component, causing those values to get pushed to the client, resetting the "current" calculation's basepoint.
* `HungerSystem.SetHunger` and anything that modifies the current threshold call `HungerSystem.SetAuthoritativeHungerValue` to ensure any unpredictable changes are replicated to the client.

## Media
*Before:*
(With non-game-logic-altering modifications made to raise an event on `HungerComponent` replication + popup creation when getting that event)

https://github.com/user-attachments/assets/f1559b39-e362-4091-8652-d84a84155338

*After:*
(With non-game-logic-altering modifications made to raise an event on `HungerComponent` replication + popup creation when getting that event; and create a popup on every `HungerSystem` update tick)

https://github.com/user-attachments/assets/9409118c-c159-4b9a-826f-f6412932a8a8

## Requirements
- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes
* `HungerComponent`'s `CurrentHunger` no longer exists and must be replaced by a call to `HungerSystem.GetHunger`

**Changelog**
My understanding is that this is a totally under-the-hood change, so no CL.

### CODE REVIEW

**Comment by Centronias** [comment]
I'll do the same thing to thirst if this PR doesn't get obliterated for my implementation.

**Comment by slarticodefast** [comment]
> I'll do the same thing to thirst if this PR doesn't get obliterated for my implementation.

Ideally someone finishes this derelict PR to combine the hunger and thirst systems into satiation #27481
But until then it should be fine to improve things step by step 👍

**Review by PJB3005 — CHANGES_REQUESTED** [review]
This is absolutely not the right way to do this. The issue you linked literally explains the correct way to fix it instead. Do that.

And do not do random style changes/refactors in unrelated PRs.

File: Content.Server/EntityEffects/EffectConditions/TotalHunger.cs [hunk 1]
```diff
@@ -15,9 +15,10 @@ public sealed partial class Hunger : EntityEffectCondition
 
     public override bool Condition(EntityEffectBaseArgs args)
     {
-        if (args.EntityManager.TryGetComponent(args.TargetEntity, out HungerComponent? hunger))
+        if (args.EntityManager.TryGetComponent(args.TargetEntity, out HungerComponent? hunger) &&
+            args.EntityManager.SystemOrNull<HungerSystem>() is { } hungerSystem)
```
Comment (Centronias) [inline Content.Server/EntityEffects/EffectConditions/TotalHunger.cs:19]: I don't know if `args.EntityManager.SystemOrNull<HungerSystem>()` is the right thing to do here, but I am not sure how else to access the system.

File: Resources/Prototypes/Entities/Mobs/NPCs/animals.yml [hunk 2]
```diff
@@ -1686,7 +1686,7 @@
       Dead: 0
     baseDecayRate: 0.04
   - type: Hunger
-    currentHunger: 25   # spawn with Okay hunger state
```
Comment (Centronias) [inline Resources/Prototypes/Entities/Mobs/NPCs/animals.yml:1689]: This name change definitely sucks, but I don't know a way around it.

It would be good for `currentHunger` or `startingHunger` to be the field in the yaml, but it would, I think, not be good to  have either of those names show up in-game in something like ViewVariables.

File: Content.Shared/Nutrition/EntitySystems/HungerSystem.cs [hunk 3]
```diff
@@ -120,7 +151,7 @@ private void UpdateCurrentThreshold(EntityUid uid, HungerComponent? component =
             return;
         component.CurrentThreshold = calculatedHungerThreshold;
         DoHungerThresholdEffects(uid, component);
-        Dirty(uid, component);
+        SetAuthoritativeHungerValue((uid, component), GetHunger(component));
```
Comment (PJB3005) [inline Content.Shared/Nutrition/EntitySystems/HungerSystem.cs:154]: Why is this necessary?

File: Resources/Prototypes/Entities/Mobs/NPCs/animals.yml (hunk 2 above)
Comment (PJB3005) [reply Resources/Prototypes/Entities/Mobs/NPCs/animals.yml:1689]: > This name change definitely sucks, but I don't know a way around it.
> 
> It would be good for `currentHunger` or `startingHunger` to be the field in the yaml, but it would, I think, not be good to have either of those names show up in-game in something like ViewVariables.

`[ViewVariables(VVAccess.ReadOnly)]` it.

Also I'm looking at the code, does setting this data field from YAML even work? It looks like it would always get replaced by the randomized amount selected in `OnMapInit`. (Note that this doesn't appear to be an issue with your PR, just an observation).

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: Content.Shared/Nutrition/EntitySystems/HungerSystem.cs (hunk 3 above)
Comment (Centronias) [reply Content.Shared/Nutrition/EntitySystems/HungerSystem.cs:154]: My thinking is any time the threshold changes, the "start point" values should be updated as well.
I suppose that's not strictly true as the only change that necessitates an update is a change to the decay rate. So I'll move this call down to exactly where the decay rate is modified.

File: Resources/Prototypes/Entities/Mobs/NPCs/animals.yml (hunk 2 above)
Comment (Centronias) [reply Resources/Prototypes/Entities/Mobs/NPCs/animals.yml:1689]: It looks like this was changed with the last rewrite of all of this:
https://github.com/space-wizards/space-station-14/pull/14939/files#diff-7ff03c169eb454aed40cb0979bd1d81f65ce0db09590d25191682c783e7ac344L131

Should I get rid of the fields in the yaml, since they don't do anything now?

File: Content.Server/EntityEffects/EffectConditions/TotalHunger.cs (hunk 1 above)
Comment (Centronias) [reply Content.Server/EntityEffects/EffectConditions/TotalHunger.cs]: The only `GetEntitySystem` I could find is on `IEntitySystemManager` which, I guess, would need to be injected, but I assume `[Dependency]`s can't be injected into non-systems like this.

Instead, I've changed it to just do `args.EntityManager.System<...>()`, based on
https://github.com/space-wizards/space-station-14/blob/fdd713a0f0e5278ef9ae24c93dcbc7c8e8094e15/Content.Server/EntityEffects/EffectConditions/HasTagCondition.cs#L21
doing that same thing.

If this is still not the right thing, do you mind providing me with a bit more guidance?

File: Content.Shared/Nutrition/EntitySystems/HungerSystem.cs [hunk 4]
```diff
@@ -23,6 +24,7 @@ public sealed class HungerSystem : EntitySystem
     [Dependency] private readonly MobStateSystem _mobState = default!;
     [Dependency] private readonly MovementSpeedModifierSystem _movementSpeedModifier = default!;
     [Dependency] private readonly SharedJetpackSystem _jetpack = default!;
+    [Dependency] private readonly INetManager _net = default!;
```
Comment (PJB3005) [inline Content.Shared/Nutrition/EntitySystems/HungerSystem.cs:27]: This is unused.

File: Resources/Prototypes/Entities/Mobs/NPCs/animals.yml (hunk 2 above)
Comment (PJB3005) [reply Resources/Prototypes/Entities/Mobs/NPCs/animals.yml:1689]: Your choice I suppose

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Review by PJB3005 — APPROVED** [review]

**Review by SlamBamActionman — APPROVED** [review]
I've tested it ingame and yeah, it seems to work as advertised. I guess next step will be to predict metabolism, eh? ;)

Thank you for your contribution!
